core/src/layers/metrics.rs

```diff
@@ -70,148 +67,112 @@ use crate::*;
 /// let recorder = builder.build_recorder().expect("failed to build recorder");
 /// ```
 #[derive(Clone, Debug, Default)]
-pub struct MetricsLayer {
-    path_label_level: usize,
-}
-
-impl MetricsLayer {
-    /// Set the level of path label.
-    ///
-    /// - level = 0: we will ignore the path label.
-    /// - level > 0: the path label will be the path split by "/" and get the last n level,
-    ///   if n=1 and input path is "abc/def/ghi", and then we will get "abc/" as the path label.
-    pub fn path_label(mut self, level: usize) -> Self {
-        self.path_label_level = level;
-        self
-    }
-}
+pub struct MetricsLayer {}
 
 impl<A: Access> Layer<A> for MetricsLayer {
     type LayeredAccess = observe::MetricsAccessor<A, MetricsInterceptor>;
 
     fn layer(&self, inner: A) -> Self::LayeredAccess {
-        let interceptor = MetricsInterceptor {
-            path_label_level: self.path_label_level,
-        };
+        let interceptor = MetricsInterceptor {};
         observe::MetricsLayer::new(interceptor).layer(inner)
     }
 }
 
 #[derive(Clone, Debug)]
-pub struct MetricsInterceptor {
-    path_label_level: usize,
-}
+pub struct MetricsInterceptor {}
 
 impl observe::MetricsIntercept for MetricsInterceptor {
-    fn observe_operation_duration_seconds(
-        &self,
-        info: Arc<AccessorInfo>,
-        path: &str,
-        op: Operation,
-        duration: Duration,
-    ) {
-        let labels = OperationLabels {
-            info,
-            path: observe::path_label_value(path, self.path_label_level),
-            operation: op,
-            error: None,
+    fn observe(&self, labels: observe::MetricLabels, value: observe::MetricValue) {
+        let labels = OperationLabels(labels).into_labels();
+
+        match value {
+            observe::MetricValue::OperationBytes(v) => {
+                histogram!(value.name(), labels).record(v as f64)
+            }
+            observe::MetricValue::OperationBytesRate(v) => {
+                histogram!(value.name(), labels).record(v)
+            }
+            observe::MetricValue::OperationEntries(v) => {
+                histogram!(value.name(), labels).record(v as f64)
+            }
+            observe::MetricValue::OperationEntriesRate(v) => {
+                histogram!(value.name(), labels).record(v)
+            }
+            observe::MetricValue::OperationDurationSeconds(v) => {
+                histogram!(value.name(), labels).record(v)
+            }
+            observe::MetricValue::OperationErrorsTotal => {
+                counter!(value.name(), labels).increment(1)
+            }
+            observe::MetricValue::OperationExecuting(v) => {
+                if v.is_positive() {
```

Comment:
This seems to be wrong:

- If `v=1`, we are `increment(1)`
- If `v=-1`, we are `decrement(-1)` => `increment(1)`.

Reply:
Yeah, it's a mistake. For metrics crate, the `GaugeFn` trait is only implemented for `AtomicU64`, so we should use `decrement(-v as f64)` here

```
pub struct Gauge {
    inner: Option<Arc<dyn GaugeFn + Send + Sync>>,
}

impl GaugeFn for AtomicU64 {
  fn decrement(&self, value: f64) {
    // ...
  }
}
```